Lib/test/test_gc.py

```diff
@@ -959,6 +964,69 @@ def getstats():
 
         gc.enable()
 
+    @unittest.skipIf(ContainerNoGC is None,
+                     'requires ContainerNoGC extension type')
+    def test_trash_weakref_clear(self):
+        # Test that trash weakrefs are properly cleared (bpo-38006).
+        #
+        # Structure we are creating:
+        #
+        #   Z <- Y <- A--+--> WZ -> C
+        #             ^  |
+        #             +--+
+        # where:
+        #   WZ is a weakref to Z with callback C
+        #   Y doesn't implement tp_traverse
+        #   A contains a reference to itself, Y and WZ
+        #
+        # A, Y, Z, WZ are all trash.  The GC doesn't know that Z is trash
+        # because Y does not implement tp_traverse.  To show the bug, WZ needs
+        # to live long enough so that Z is deallocated before it.  Then, if
+        # gcmodule is buggy, when Z is being deallocated, C will run.
+        #
+        # To ensure WZ lives long enough, we put it in a second reference
```

Comment:
Idea: Another thing we can do to make sure WZ lives long enough is having some object in the same cycle as A that resurrects, resurrecting WZ with it. In this way, we know that WZ will be alive when `delete_garbage` is executed.

Reply:
Except it's not enough that WZ merely be alive, it must also be that WZ hasn't been cleared (in the weakref sense, not the `tp_clear` sense).  But Neil fixed this problem by ensuring WZ _is_ cleared, long before we get anywhere near `delete_garbage`.  So deallocating Z will no longer trigger C.

So the problem can't be triggered in code since Neil's fix.  Code before that is also before the more-recent finalizer changes, so in those running a finalizer that resurrects something causes the gc run to abort, before  `delete_garbage` is run.  So Z, WZ, and C would all survive then, but undamaged.

Reply:
> Except it's not enough that WZ merely be alive, it must also be that WZ hasn't been cleared (in the weakref sense, not the `tp_clear` sense)

But if I deactivate Neil's fix for this problem, with the current state of the gc, the scenario that I described should trigger the error (the callback is invoked), no?

Reply:
Probably, but we really want a test that fails under some version of Python that was actually released.

Reply:
> Probably, but we really want a test that fails under some version of Python that was actually released.

Oh, my bad then. I thought this test was not going to be backported.

Reply:
I don't think backporting is relevant.  To show that a bug has been fixed, we need a test case that fails under an earlier version.  There's no need to backport it.  It's more:  "here:  run this under 3.7.2 (whatever).  See?  It fails.  But under current Python it doesn't.".  Indeed, we never backport tests to versions under which they fail 😉

Reply:
> Indeed, we never backport tests to versions under which they fail

Gotcha, everything clear. For one second I was thinking that we were going to backport the test *and* the fix. :)

Reply:
We can backport the test and the fix (and should!), but we certainly can't include the recent gc finalizer enhancements in old-version bugfix releases.

Reply:
If we had something like ContainerNoGC in old Python versions, the test would be easy to backport and should fail in versions going back to 2.7 (and earlier, up to when weakrefs were added).  If you want to make Python crash, find an object where tp_clear leaves it in an invalid state.   types.SimpleNamespace is one example.  Probably nicer if the unit tests don't SEGV on failure. ;-P